pmlc/rt/vulkan/vulkan_runtime.cc

```diff
@@ -160,25 +304,38 @@ LogicalResult VulkanRuntime::run() {
   if (failed(createDescriptorSetLayout()) || failed(createPipelineLayout()) ||
       // Each descriptor set must be allocated from a descriptor pool.
       failed(createComputePipeline()) || failed(createDescriptorPool()) ||
-      failed(allocateDescriptorSets()) || failed(setWriteDescriptors()) ||
-      // Create command buffer.
-      failed(createCommandPool()) || failed(createComputeCommandBuffer())) {
+      failed(allocateDescriptorSets()) || failed(setWriteDescriptors())) {
     return failure();
   }
 
-  // Get working queue.
-  vkGetDeviceQueue(device, queueFamilyIndex, 0, &queue);
+  return success();
+}
 
+void VulkanRuntime::addLaunchActionToSchedule() {
+  schedule.push_back(curr);
+  curr = nullptr;
+}
+
+LogicalResult VulkanRuntime::submitCommandBuffers() {
+  if (failed(createSchedule())) {
+    return failure();
+  }
   // Submit command buffer into the queue.
-  if (failed(submitCommandBuffersToQueue()))
+  if (failed(submitCommandBuffersToQueue())) {
     return failure();
+  }
 
   RETURN_ON_VULKAN_ERROR(vkQueueWaitIdle(queue), "vkQueueWaitIdle");
+
+  updateHostMemoryBuffers();
   return success();
 }
 
 LogicalResult VulkanRuntime::createInstance() {
-  RETURN_ON_VULKAN_ERROR(volkInitialize(), "volkInitialize");
+  if (volkInitialized != VK_SUCCESS) {
+    volkInitialized = volkInitialize();
+    RETURN_ON_VULKAN_ERROR(volkInitialized, "volkInitialize");
+  }
```

Comment:
Why do we call `createInstance` more than once? That seems like the real issue here.

Reply:
Yes, I am still WIP to make the changes per our discussion earlier this week to split the main into init and main so vulkan will be initialized once and not destroyed every run.